state-chain/pallets/cf-ingress-egress/src/boost_pool.rs

```diff
@@ -0,0 +1,281 @@
+		self.available_amount.saturating_accrue(added_amount);
+	}
+
+	pub(crate) fn add_funds(&mut self, account_id: AccountId, added_amount: C::ChainAmount) {
+		self.add_funds_inner(account_id, ScaledAmount::from_chain_amount(added_amount));
+	}
+
+	pub(crate) fn get_available_amount(&self) -> C::ChainAmount {
+		self.available_amount.into_chain_amount()
+	}
+
+	/// Records `amount_needed` as being used for boosting and to be re-distributed
+	/// among current boosters (along with the fee) upon finalisation
+	pub(crate) fn use_funds_for_boosting(
+		&mut self,
+		boost_id: BoostId,
+		amount_needed: C::ChainAmount,
+		boost_fee: C::ChainAmount,
+	) -> Result<(), &'static str> {
+		let amount_needed = ScaledAmount::from_chain_amount(amount_needed);
+		let fees = ScaledAmount::from_chain_amount(boost_fee);
+
+		let current_total_active_amount = self.available_amount;
+
+		self.available_amount = self
+			.available_amount
+			.checked_sub(amount_needed)
+			.ok_or("Not enough active funds")?;
+
+		let mut amounts = self.amounts.iter_mut();
+
+		// We must have at least one entry because the available amount is non-zero:
+		let first_entry = amounts.next().ok_or("No boost amount entries found")?;
+
+		let mut total_contributed = ScaledAmount::<C>::default();
+		let mut to_receive_recorded = ScaledAmount::default();
+
+		let amount_to_receive = amount_needed.saturating_add(fees);
+
+		let mut boosters_to_receive: BTreeMap<_, _> = amounts
+			.map(|(booster_id, amount)| {
+				let booster_contribution = multiply_by_rational_with_rounding(
+					amount_needed.into(),
+					(*amount).into(),
+					current_total_active_amount.into(),
+					Rounding::NearestPrefDown,
+				)
+				// booster's amount is always <= total amount so default due to overflow should be
+				// impossible
+				.unwrap_or_default()
+				.into();
+
+				// Same as above, but also includes fees:
+				let booster_to_receive = multiply_by_rational_with_rounding(
+					amount_to_receive.into(),
+					(*amount).into(),
+					current_total_active_amount.into(),
+					Rounding::NearestPrefDown,
+				)
+				// booster's amount is always <= total amount so default due to overflow should be
+				// impossible
+				.unwrap_or_default()
+				.into();
+
+				// Amount should always be large enough at this point, but saturating to be safe:
+				amount.saturating_reduce(booster_contribution);
+
+				total_contributed.saturating_accrue(booster_contribution);
+				to_receive_recorded.saturating_accrue(booster_to_receive);
+
+				(booster_id.clone(), booster_to_receive)
+			})
+			.collect();
+
+		// Compute the amount for the first contributor working backwards from the amount needed
+		// to negate any rounding errors made so far:
+		let remaining_required = amount_needed.saturating_sub(total_contributed);
+		let remaining_to_receive = amount_to_receive.saturating_sub(to_receive_recorded);
+		first_entry.1.saturating_reduce(remaining_required);
```

Comment:
You are right, it is possible (though very unlikley), and we should probably avoid that. I think rounding up is also problematic since we might end up in a situation where `n-1` boosters provide more than needed, and then we might end up crediting them the funds that we dont have.

What if we round up when deducting from boosters' amounts, but round down when calculating amounts owed to them?

Reply:
Updated the code now to correctly account for any rounding errors.